infra/Pos/Diffusion/Subscription/Dns.hs

```diff
@@ -54,52 +60,57 @@ dnsSubscriptionWorker oq networkCfg DnsDomains{..} keepaliveTimer nextSlotDurati
     -- fallbacks (for a given outer list element) is the length of the inner
     -- list (disjuncts).
     logNotice $ sformat ("dnsSubscriptionWorker: valency "%int) (length allOf)
-    void $ forConcurrently allOf (subscribeAlts dnsPeersVar)
+    void $ forConcurrently allOf (subscribeAlts dnsPeersVar subDuration)
     logNotice $ sformat ("dnsSubscriptionWorker: all "%int%" threads finished") (length allOf)
   where
 
     allOf :: [(Int, Alts (NodeAddr DNS.Domain))]
     allOf = zip [1..] dnsDomains
 
-    -- Resolve all of the names and try to subscribe to one.
+    -- Resolve all of the namtas and try to subscribe to one.
     -- If a subscription goes down, try later names.
     -- When the list is exhausted (either because it's empty to begin with, or
     -- because all subscriptions to have failed), wait a while before retrying
     -- (see 'retryInterval').
     subscribeAlts
         :: SharedAtomicT m (Map Int (Alts NodeId))
+        -> MVar Millisecond
         -> (Int, Alts (NodeAddr DNS.Domain))
         -> m ()
-    subscribeAlts _ (index, []) =
+    subscribeAlts _ _ (index, []) =
         logWarning $ sformat ("dnsSubscriptionWorker: no alternatives given for index "%int) index
-    subscribeAlts dnsPeersVar (index, alts) = do
+    subscribeAlts dnsPeersVar subDuration (index, alts) = do
         -- Any DNSError is squelched. So are IOExceptions, for good measure.
         -- This does not include async exceptions.
         -- It does handle the case in which there's no internet connection, or
         -- a bad configuration, so that the subscription thread will keep on
         -- retrying.
-        findAndSubscribe dnsPeersVar index alts
+        findAndSubscribe dnsPeersVar subDuration index alts
             `catch` logDNSError
             `catch` logIOException
-        retryInterval >>= delay
-        subscribeAlts dnsPeersVar (index, alts)
+        d <- swapMVar subDuration 0 >>= retryInterval
```

Comment:
Why `swapMVar` to `0`? Didn't we take a `max` before?

Reply:
Here at this point we tried to subscribe to every body so we can zero the counter and measure the max time for the next round.  This way we don't accumulate maximal time if something worked well and then it stopped.

We compute max time for each round, i.e. per `subscribeAlts` call. The `max` you are referring to is [here](https://github.com/input-output-hk/cardano-sl/pull/2544/files#diff-555a6bf9a7a1fb7e3438987fd51919aeR86).

Reply:
Ah but the `subDuration` `MVar` is shared between potentially more than 1 thread. It's a conjunction of disjunctions. `subscribeAlts` runs through a single conjunct, but there could be others.

Reply:
You are right, this is cleared to often. Instead we should clear it once the outer list (conjuctions) are consumed.

Reply:
After a pause of thinking, I think it is just better to create the `subDuration` `MVar` in each thread.  This way it will control how fast `subscribeAlts` spins through a list of alternatives.

Reply:
Yes that sounds reasonable. Also `subscribeAlts` should just take an `MVar Millisecond` rather than a `Maybe (MVar Millisecond)`.

Reply:
So you prefer that the context provides it, here the `MVar` is created by `subscribeAlts`. But ok, let me changes this, it will be more intuitive.